File: Loenn/entities/station_blocks/station_block_track.lua
local drawableSprite = require("structs.drawable_sprite")
local utils = require("utils")

local stationBlockTrack = {}

local switchStates = {"None", "On", "Off"}
local moveModes = {"None", "ForwardOneWay", "BackwardOneWay", "ForwardForce", "BackwardForce"}

stationBlockTrack.name = "CommunalHelper/StationBlockTrack"
stationBlockTrack.depth = -5000
stationBlockTrack.minimumSize = {24, 24}
stationBlockTrack.canResize = {true, true}
stationBlockTrack.fieldInformation = {
    indicatorColor = {
        fieldType = "color"
    },
    indicatorIncomingColor = {
        fieldType = "color"
    }
}

stationBlockTrack.placements = {}
for i, state in ipairs(switchStates) do
    stationBlockTrack.placements[2 * i - 1] = {
        name = "horiz_" .. string.lower(state),
        placementType = "rectangle",
        data = {
            width = 24,
            height = 8,
            horizontal = true,
            trackSwitchState = state,
            moveMode = "None",
            multiBlockTrack = false,
            indicator = true,
            indicatorColor = "008080",
            indicatorIncomingColor = "c92828"
        }
    }
    stationBlockTrack.placements[2 * i] = {
        name = "verti_" .. string.lower(state),
        placementType = "rectangle",
        data = {
            width = 8,
            height = 24,
            horizontal = false,
            trackSwitchState = state,
            moveMode = "None",
            multiBlockTrack = false,
            indicator = true,
            indicatorColor = "008080",
            indicatorIncomingColor = "c92828"
        }
    }
end

local noneColor = {1.0, 1.0, 1.0, 1.0}
local onColor = {66 / 255, 167 / 255, 1.0, 1.0}
local offColor = {1.0, 48 / 255, 131 / 255, 1.0}

local nodesTexture = "objects/CommunalHelper/stationBlock/tracks/outline/node"
local hTrackTexture = "objects/CommunalHelper/stationBlock/tracks/outline/h"
local vTrackTexture = "objects/CommunalHelper/stationBlock/tracks/outline/v"
local arrowsTexture = "objects/CommunalHelper/stationBlock/tracks/outline/arrows"

local function addMoveModeIndicatorSprites(sprites, entity, width, height, moveMode, ty, ox, oy, color)
    local force = moveMode == "backwardforce" or moveMode == "forwardforce"
    local backward = moveMode == "backwardoneway" or moveMode == "backwardforce"

    local stop = force and 1 or 0
    for i = 0, stop do
        local offset = 3 * i * (backward and -1 or 1)

        local arrowSprite = drawableSprite.fromTexture(arrowsTexture, entity)
        arrowSprite:setJustification(0.0, 0.0)
        arrowSprite:useRelativeQuad(backward and 0 or 8, ty, 8, 8)
        arrowSprite:addPosition(ox * ((backward and -6 or (width - 2)) + offset), oy * ((backward and -6 or (height - 2)) + offset))
        arrowSprite:setColor(color)

        table.insert(sprites, arrowSprite)
    end
end 

local function addTrackSprites(sprites, entity, horiz, color, width, height, ty, ox, oy)
    local nodeFromSprite = drawableSprite.fromTexture(nodesTexture, entity)
    nodeFromSprite:setColor(color)
    nodeFromSprite:setJustification(0.0, 0.0)
    nodeFromSprite:useRelativeQuad(0, ty, 8, 8)

    local nodeToSprite = drawableSprite.fromTexture(nodesTexture, entity)
    nodeToSprite:addPosition(ox * (width - 8), oy * (height - 8))
    nodeToSprite:setColor(color)
    nodeToSprite:setJustification(0.0, 0.0)
    nodeToSprite:useRelativeQuad(8, ty, 8, 8)

    table.insert(sprites, nodeFromSprite)
    table.insert(sprites, nodeToSprite)
    
    local size = math.floor((horiz and width or height) / 8)
    local trackTexture = horiz and hTrackTexture or vTrackTexture

    for i = 1, size - 2 do
        local trackSprite = drawableSprite.fromTexture(trackTexture, entity)
        trackSprite:setColor(color)
        trackSprite:setJustification(0.0, 0.0)
        trackSprite:addPosition(ox * i * 8, oy * i * 8)

        table.insert(sprites, trackSprite)
    end
end

function stationBlockTrack.sprite(room, entity)
    local sprites = {}

    local width, height = entity.width or 24, entity.height or 24
    local horiz = entity.horizontal
    local ox, oy = horiz and 1 or 0, horiz and 0 or 1
    local ty = horiz and 0 or 8

    local state = string.lower(entity.trackSwitchState or "None")
    local color = ((state == "on") and onColor) or ((state == "off") and offColor) or noneColor

    local moveMode = string.lower(entity.moveMode or "None")

    if moveMode ~= "none" then
        addMoveModeIndicatorSprites(sprites, entity, width, height, moveMode, ty, ox, oy, color)
    end

    addTrackSprites(sprites, entity, horiz, color, width, height, ty, ox, oy)

    return sprites
end

function stationBlockTrack.rectangle(room, entity)
    if entity.horizontal then
        return utils.rectangle(entity.x, entity.y, entity.width or 24, 8)
    end
    return utils.rectangle(entity.x, entity.y, 8, entity.height or 24)
end

return stationBlockTrack

File: Ahorn/entities/misc/StationBlockTrack.jl
module CommunalHelperStationBlockTrack

using ..Ahorn, Maple

const switchStates = ["None", "On", "Off"]
const moveModes = ["None", "ForwardOneWay", "BackwardOneWay", "ForwardForce", "BackwardForce"]

@mapdef Entity "CommunalHelper/StationBlockTrack" StationBlockTrack(
    x::Integer,
    y::Integer,
    width::Integer=24,
    height::Integer=24,
    horizontal::Bool=false,
    trackSwitchState::String="None",
    moveMode::String="None",
    multiBlockTrack::Bool=false,
    dynamicRouting::Bool=false,
    indicator::Bool=true,
    indicatorColor::String="008080",
    indicatorIncomingColor::String="c92828",
)

const placements = Ahorn.PlacementDict(
    "Station Block Track ($orientation, $stateLabel) (Communal Helper)" => Ahorn.EntityPlacement(
        StationBlockTrack,
        "rectangle",
        Dict{String,Any}(
            "horizontal" => orientation == "Horizontal",
            "trackSwitchState" => state,
        ),
    ) for orientation in ["Vertical", "Horizontal"], (state, stateLabel) in zip(switchStates, ["No Switching", "Switch On", "Switch Off"])
)

Ahorn.minimumSize(entity::StationBlockTrack) = 24, 24

function Ahorn.resizable(entity::StationBlockTrack) 
    horiz = Bool(get(entity.data, "horizontal", false))
    return (horiz, !horiz)
end

Ahorn.editingOptions(entity::StationBlockTrack) = Dict{String,Any}(
    "trackSwitchState" => switchStates,
    "moveMode" => moveModes,
)

# very weird rotation
function Ahorn.rotated(entity::StationBlockTrack, steps::Int)
    horiz = Bool(get(entity.data, "horizontal", false))
    trackSwitchState = get(entity.data, "trackSwitchState", "None")
    return StationBlockTrack(entity.x, entity.y, entity.height, entity.width, !horiz, trackSwitchState)
end

function Ahorn.selection(entity::StationBlockTrack)
    x, y = Ahorn.position(entity)

    width = Int(get(entity.data, "width", 24))
    height = Int(get(entity.data, "height", 24))
    horiz = Bool(get(entity.data, "horizontal", false))

    return horiz ? Ahorn.Rectangle(x, y, width, 8) : Ahorn.Rectangle(x, y, 8, height)
end

const nodeSprite = "objects/CommunalHelper/stationBlock/tracks/outline/node"
const hTrack = "objects/CommunalHelper/stationBlock/tracks/outline/h"
const vTrack = "objects/CommunalHelper/stationBlock/tracks/outline/v"
const arrows = "objects/CommunalHelper/stationBlock/tracks/outline/arrows"

const noneColor = (255, 255, 255, 255) ./ 255
const onColor = (66, 167, 255, 255) ./ 255
const offColor = (255, 48, 131, 255) ./ 255

function getTrackColor(entity::StationBlockTrack)
    trackSwitchState = get(entity.data, "trackSwitchState", "None")
    if trackSwitchState == "On"
        return onColor
    elseif trackSwitchState == "Off"
        return offColor
    else
        return noneColor
    end
end

function Ahorn.renderAbs(ctx::Ahorn.Cairo.CairoContext, entity::StationBlockTrack)
    x, y = Ahorn.position(entity)

    width = Int(get(entity.data, "width", 24))
    height = Int(get(entity.data, "height", 24))

    horiz = Bool(get(entity.data, "horizontal", false))

    color = getTrackColor(entity)

    moveMode = String(get(entity.data, "moveMode", "None"))
    backwardArrowCount = moveMode == "BackwardForce" ? 2 : (moveMode == "BackwardOneWay" ? 1 : 0)
    forwardArrowCount = moveMode == "ForwardForce" ? 2 : (moveMode == "ForwardOneWay" ? 1 : 0)

    if horiz
        tilesWidth = div(width, 8)

        for i in 2:tilesWidth-1
            Ahorn.drawImage(ctx, hTrack, x + (i - 1) * 8, y, tint=color)
        end

        Ahorn.drawImage(ctx, nodeSprite, x, y, 0, 0, 8, 8, tint=color)
        Ahorn.drawImage(ctx, nodeSprite, x + width - 8, y, 8, 0, 8, 8, tint=color)
        for i in 0:backwardArrowCount-1
            Ahorn.drawImage(ctx, arrows, x - 6 - 3 * i, y, 0, 0, 8, 8, tint=color)
        end
        for i in 0:forwardArrowCount-1
            Ahorn.drawImage(ctx, arrows, x + width - 2 + 3 * i, y, 8, 0, 8, 8, tint=color)
        end
    else
        tilesHeight = div(height, 8)

        for i in 2:tilesHeight-1
            Ahorn.drawImage(ctx, vTrack, x, y + (i - 1) * 8, tint=color)
        end

        Ahorn.drawImage(ctx, nodeSprite, x, y, 8, 8, 8, 8, tint=color)
        Ahorn.drawImage(ctx, nodeSprite, x, y + height - 8, 0, 8, 8, 8, tint=color)
        for i in 0:backwardArrowCount-1
            Ahorn.drawImage(ctx, arrows, x, y - 6 - i * 3, 0, 8, 8, 8, tint=color)
        end
        for i in 0:forwardArrowCount-1
            Ahorn.drawImage(ctx, arrows, x, y + height - 2 + i * 3, 8, 8, 8, 8, tint=color)
        end
    end
end

end
